src/modules/ekf2/EKF/gps_control.cpp

```diff
@@ -46,187 +46,102 @@ void Ekf::controlGpsFusion(const imuSample &imu_delayed)
 		return;
 	}
 
+	if (!gyro_bias_inhibited()) {
+		_yawEstimator.setGyroBias(getGyroBias());
+	}
+
+	// run EKF-GSF yaw estimator once per imu_delayed update
+	_yawEstimator.predict(imu_delayed, _control_status.flags.in_air && !_control_status.flags.vehicle_at_rest);
+
 	_gps_intermittent = !isNewestSampleRecent(_time_last_gps_buffer_push, 2 * GNSS_MAX_INTERVAL);
 
 	// check for arrival of new sensor data at the fusion time horizon
 	_gps_data_ready = _gps_buffer->pop_first_older_than(imu_delayed.time_us, &_gps_sample_delayed);
 
 	if (_gps_data_ready) {
-		// correct velocity for offset relative to IMU
-		const Vector3f pos_offset_body = _params.gps_pos_body - _params.imu_pos_body;
-		const Vector3f vel_offset_body = _ang_rate_delayed_raw % pos_offset_body;
-		const Vector3f vel_offset_earth = _R_to_earth * vel_offset_body;
-		_gps_sample_delayed.vel -= vel_offset_earth;
-
-		// correct position and height for offset relative to IMU
-		const Vector3f pos_offset_earth = _R_to_earth * pos_offset_body;
-		_gps_sample_delayed.pos -= pos_offset_earth.xy();
-		_gps_sample_delayed.hgt += pos_offset_earth(2);
-
-		// update GSF yaw estimator velocity (basic sanity check on GNSS velocity data)
-		if ((_gps_sample_delayed.sacc > 0.f) && (_gps_sample_delayed.sacc < _params.req_sacc)
-		    && _gps_sample_delayed.vel.isAllFinite()
-		   ) {
-			_yawEstimator.setVelocity(_gps_sample_delayed.vel.xy(), math::max(_gps_sample_delayed.sacc, _params.gps_vel_noise));
-		}
-	}
+		if  (runGnssChecks(_gps_sample_delayed) && isTimedOut(_last_gps_fail_us, (uint64_t)_min_gps_health_time_us / 2)) {
+			if (isTimedOut(_last_gps_fail_us, (uint64_t)_min_gps_health_time_us)) {
+				// First time checks are passing, latching.
+				_gps_checks_passed = true;
+			}
 
-	if (!gyro_bias_inhibited()) {
-		_yawEstimator.setGyroBias(getGyroBias());
-	}
+			collect_gps(_gps_sample_delayed);
 
-	// run EKF-GSF yaw estimator once per imu_delayed update
-	_yawEstimator.update(imu_delayed, _control_status.flags.in_air && !_control_status.flags.vehicle_at_rest);
+		} else {
+			// Skip this sample
+			_gps_data_ready = false;
 
-	// Check for new GPS data that has fallen behind the fusion time horizon
-	if (_gps_data_ready) {
+			if (_control_status.flags.gps && isTimedOut(_last_gps_pass_us, _params.reset_timeout_max)) {
+				stopGpsFusion();
+				_warning_events.flags.gps_quality_poor = true;
+				ECL_WARN("GPS quality poor - stopping use");
+			}
+		}
```

Comment:
Here maybe? Needs to go after the origin is potentially set by collect_gps().

```suggestion
		}
		
		updateGnssVel(_gps_sample_delayed, _aid_src_gnss_vel);
		updateGnssPos(_gps_sample_delayed, _aid_src_gnss_pos);
```

Reply:
Done